blog/2025-01-13-typescript-sdk-release.md

```diff
@@ -0,0 +1,375 @@
+- In particular instances, the `$type` property can actually have two values to describe the same lexicon. An "images" embed, for example, can use both `app.bsky.embed.images` and `app.bsky.embed.images#main` as `$type`. This makes the previous point even worse.
+
+In order to alleviate these issues, the SDK provides type checking predicate functions. In their previous implementation, the `is*` utilities were defined as follows:
+
+```typescript
+export interface Main {
+  images: Image[]
+  [x: string]: unknown
+}
+
+export function isMain(value: unknown): values is Main {
+  return (
+    value != null &&
+    typeof value === 'object' &&
+    '$type' in value &&
+    (value.$type === 'app.bsky.embed.images' ||
+      value.$type === 'app.bsky.embed.images#main')
+  )
+}
+```
+
+As can be seen from the example implementation above, the predicate functions would cast any object containing the expected `$type` property into the corresponding type, without checking for the actual validity of other properties. This could yield runtime errors that could have been avoided during development:
+
+```typescript
+import { AppBskyEmbedImages } from '@atproto/api'
+
+// Alias, for clarity
+const isImages = AppBskyEmbedImages.isMain
+
+// Get an invalid embed somehow
+const invalidEmbed = {
+  $type: 'app.bsky.embed.images',
+  // notice how the `images` property is missing here
+}
+
+// This predicate function only checks the value of the `$type` property, making the condition "true" here
+if (isImages(invalidEmbed)) {
+  // No TypeScript error, BUT causes a runtime error because there is no "images" property !
+  console.log('First image:', invalidEmbed.images[0])
+}
+```
+
+The root of the issue here is that the `is*` utility methods perform type casting of objects solely based on the value of their `$type` property. There were basically two ways of fixing this issue:
+
+1. Alter the implementation to actually validate the object's structure. This is a non-breaking change that has a negative impact on performance.
+2. Alter the function signature to describe what the function actually does. This is a breaking change because TypeScript would start (rightfully) returning lots of errors in places where these functions are used.
+
+Because this release introduces other breaking changes, and because adapting our own codebase to this change showed it made more sense, we decided to adopt the latter option.
+
+In lots of cases where data needs to be discriminated, this change in the signature of the `is*` function will actually not cause any issue when upgrading the version of the SDK. This is the case for example when working with data obtained from the API. Because an API is a "contract" between a server and a client, the data returned by the server is "guaranteed" to be valid. In these cases, the `is*` utility methods provide a convenient way to discriminate between valid values.
```

Comment:
This note that we guarantee server response shape should be more prominent imo. Do we have admonitions set up here? If nothing else, just bolded.

Devs migrating to this _need_ to know this, otherwise they're going to think they need to use the `isValid*` method variants.